torch/csrc/jit/codegen/cuda/lower_validation.cpp

```diff
@@ -518,85 +521,214 @@ void validateParallelize(Fusion* fusion) {
+              // i.e. if producer is parallelized across threadIdx.x in a
+              // certain split, if the consumer doesn't map to this split, then
+              // we need to assume it has to be in smem with proper syncs.
+            } else {
+              auto it = std::find_if(
+                  producer->domain()->domain().begin(),
+                  producer->domain()->domain().end(),
+                  [&](IterDomain* p_id) {
+                    return loop_map.areMapped(p_id, c_id);
+                  });
+              if (it == producer->domain()->domain().end()) {
+                // Can't infer anything if producer doesn't have a matching axis
+                // to parallel consumer dim.
+                continue;
+              }
+              p_id = *it;
+            }
+          }
+
+          // Comm pattern options (when parallel types don't have matching
+          // axes) and required memory, Chart is producer parallel type,
+          // consumer parallel type Parallel types are Serial(S),
+          // threadIdx(T), blockIdx(B), Memory required for the producer is
+          // Local(L), Shared(S), Global(G), Sync is None (N/A), blockSync(B),
+          // grid_sync(G)
+          //
+          // P    C   Mem Req   Sync Type
+          // S    S      L          N/A
+          // S    T      L          N/A
+          // S    B      L          N/A
+          // T    S      S           B
+          // T    T      S           B
+          // T    B      S           B
+          // B    S      G           G
+          // B    T      G           G
+          // B    B      G           G
+
+          auto producer_ptype =
+              par_map.getConcreteMappedID(p_id)->getParallelType();
+          auto consumer_ptype = c_id == nullptr
+              ? ParallelType::Serial
+              : par_map.getConcreteMappedID(c_id)->getParallelType();
+
+          // Threads won't redundantly broadcast for shared memory outputs, so
+          // if producer is broadcast and mapped consumer is bound to threads
+          // there needs to be a raw sync for producer
+          if (p_id->isBroadcast() && !isParallelTypeThread(producer_ptype) &&
+              isParallelTypeThread(consumer_ptype) &&
+              producer->getMemoryType() == MemoryType::Shared &&
+              GpuLower::current()->concretizedBroadcastDomains().isConcretized(
+                  p_id)) {
+            raw_dims.set(consumer_ptype);
+          }
+
+          // When the producer axis is a broadcast, it is not really
+          // parallelized unless thread-predicated and concretized
+          if (isParallelTypeThread(producer_ptype) && p_id->isBroadcast() &&
+              (!parallel_bcast_doms.get(producer_ptype) ||
+               !GpuLower::current()
+                    ->concretizedBroadcastDomains()
+                    .isConcretized(p_id))) {
+            continue;
+          }
```

Comment:
Is there any specific reason this is placed here? Can it be moved forward as is the case for the consumer ID?

Reply:
Sorry, there were a lot of versions I went through, please post this comment again if you think it's still not clean.